Code/Framework/AzCore/AzCore/Metrics/EventLoggerFactoryImpl.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "EventLoggerFactoryImpl.h"
+#include <AzCore/Metrics/IEventLogger.h>
+#include <AzCore/std/functional.h>
+#include <AzCore/std/parallel/scoped_lock.h>
+#include <AzCore/std/ranges/ranges_algorithm.h>
+
+namespace AZ::Metrics
+{
+
+    void EventLoggerFactoryImpl::EventLoggerDeleter::operator()(IEventLogger* ptr) const
+    {
+        if (m_delete)
+        {
+            delete ptr;
+        }
+    }
+
+    EventLoggerFactoryImpl::EventLoggerFactoryImpl() = default;
+    EventLoggerFactoryImpl::~EventLoggerFactoryImpl() = default;
+
+    void EventLoggerFactoryImpl::VisitEventLoggers(const VisitEventLoggerInterfaceCallback& callback) const
+    {
+        auto VisitInterface = [&callback](const EventLoggerPtr& eventLogger)
+        {
+            return eventLogger != nullptr ? callback(*eventLogger) : true;
+        };
+
+        AZStd::scoped_lock lock(m_eventLoggerMutex);
+        AZStd::ranges::all_of(m_eventLoggers, VisitInterface);
+    }
+
+    AZ::Outcome<void, AZStd::unique_ptr<IEventLogger>> EventLoggerFactoryImpl::RegisterEventLogger(EventLoggerId eventLoggerId, AZStd::unique_ptr<IEventLogger> eventLogger)
+    {
+        // Transfer ownership of the event logger to the EventLoggerPtr
+        // If registration fails, ownership is transferred back to the input param   
```

Comment:
This isn't exactly true.  If registration fails, it doesn't place it back in the input param, but puts it in the `AZ::Failure` of the `Outcome`.